src/c_string.rs

```diff
@@ -0,0 +1,461 @@
+//! A fixed capacity [`CString`](https://doc.rust-lang.org/std/ffi/struct.CString.html).
+
+use crate::{vec::Vec, CapacityError};
+use core::{
+    ffi::{c_char, CStr},
+    ops::Deref,
+};
+
+/// A fixed capacity [`CString`](https://doc.rust-lang.org/std/ffi/struct.CString.html).
+///
+/// It stores up to N-1 elements with a byte reserved for the terminating null byte.
+#[derive(Clone)]
+pub struct CString<const N: usize> {
+    vec: Vec<u8, N>,
+}
+
+/// Naive implementation for `memchr`.
+///
+/// The naive implementation is somewhat competitive to libc's `memchr` or BurntSushi's optimized
+/// implementation for tiny slices (as shown by https://gist.github.com/Alexhuszagh/f9929e7d8e0277aa1281f511a841a167),
+/// which should be the average slice size for this crate due to its use case.
+/// But ideally we'd use at least BurntSushi's fallback implementation here.
+fn memchr(needle: u8, haystack: &[u8]) -> Option<usize> {
+    haystack.iter().position(|&b| b == needle)
+}
+
+impl<const N: usize> CString<N> {
+    /// Constructs a new, "empty" `CString`.
+    ///
+    /// Stores the first null byte as the first.
+    ///
+    /// ```rust
+    /// use heapless::CString;
+    /// use std::ffi::CStr;
+    ///
+    /// // Fixed-size CString that can store up to 10 elements
+    /// // (counting the null terminator)
+    /// let empty = CString::<10>::new();
+    ///
+    /// assert_eq!(empty.as_c_str(), <&CStr>::default());
+    /// assert_eq!(empty.to_str(), Ok(""));
+    /// ```
+    pub fn new() -> Self {
+        // TODO Resolve this.
+        // crate::sealed::greater_than_0::<CAP>();
```

Comment:
Can we return `Result<Self, CapacityError>`? Is there any conventional way for static capacity check in this crate?